src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ChainPal.cs

```diff
@@ -44,35 +45,70 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
             }
 
             TimeSpan remainingDownloadTime = timeout;
-            var leaf = new X509Certificate2(cert.Handle);
-            var downloaded = new HashSet<X509Certificate2>();
-            var systemTrusted = new HashSet<X509Certificate2>();
-
-            HashSet<X509Certificate2> candidates = OpenSslX509ChainProcessor.FindCandidates(
-                leaf,
-                extraStore,
-                downloaded,
-                systemTrusted,
-                ref remainingDownloadTime);
-
-            IChainPal chain = OpenSslX509ChainProcessor.BuildChain(
-                leaf,
-                candidates,
-                downloaded,
-                systemTrusted,
-                applicationPolicy,
-                certificatePolicy,
-                revocationMode,
-                revocationFlag,
-                verificationTime,
-                ref remainingDownloadTime);
-
-            if (chain.ChainStatus.Length == 0 && downloaded.Count > 0)
+
+            using (var leaf = new X509Certificate2(cert.Handle))
             {
-                SaveIntermediateCertificates(chain.ChainElements, downloaded);
-            }
+                var downloaded = new HashSet<X509Certificate2>();
+                var systemTrusted = new HashSet<X509Certificate2>();
+
+                HashSet<X509Certificate2> candidates = OpenSslX509ChainProcessor.FindCandidates(
+                    leaf,
+                    extraStore,
+                    downloaded,
+                    systemTrusted,
+                    ref remainingDownloadTime);
+
+                IChainPal chain = OpenSslX509ChainProcessor.BuildChain(
+                    leaf,
+                    candidates,
+                    downloaded,
+                    systemTrusted,
+                    applicationPolicy,
+                    certificatePolicy,
+                    revocationMode,
+                    revocationFlag,
+                    verificationTime,
+                    ref remainingDownloadTime);
+
+#if DEBUG
+                if (chain.ChainElements.Length > 0)
+                {
+                    X509Certificate2 reportedLeaf = chain.ChainElements[0].Certificate;
+                    Debug.Assert(reportedLeaf != null, "reportedLeaf != null");
+                    Debug.Assert(reportedLeaf.Equals(leaf), "reportedLeaf.Equals(leaf)");
+                    Debug.Assert(!ReferenceEquals(reportedLeaf, leaf), "!ReferenceEquals(reportedLeaf, leaf)");
+                }
+#endif
+
+                if (chain.ChainStatus.Length == 0 && downloaded.Count > 0)
+                {
+                    SaveIntermediateCertificates(chain.ChainElements, downloaded);
+                }
+
+                // Everything we put into the chain has been cloned, dispose all the originals.
+                systemTrusted.DisposeAndClear();
```

Comment:
Perhaps we could call this `DisposeAllAndClear` to make it clearer that multiple things are disposed. Also, why do we need to clear these temporary sets?